### ISSUE TITLE
Change is emitted after Store's action handler even when state is not changed

### ISSUE THREAD

**Issue by endragor** [goatslacker/alt#319]
It looks like alt code emits change event from Stores even when nothing was actually changed, i.e. when `setState` was not even called in the handler. I wonder what's the reason behind such behaviour, if it's intended (it seems that it is, looking at the code).
I have cases when `setState` might not be called from the action handler under certain circumstances. Besides, `setState` might be called to set a primitive value (boolean, number or string) that already has the same value (for example, boolean flag was already set to `true`). To prevent this, I have to manually insert `this.preventDefault()` and manually check whether my `setState()` is actually going to change something to save performance. I would love if the framework did it for me. I know it might be not reasonable to always compare the whole new state and old state, but just comparing the primitive values would already cover many cases, in my opinion.

**Comment by mercuriete** [comment]
whenever you return something different to false then the store emit a change.

i had this problem before.
if you debug the handler and make some steps after the return you will see that alt emit a change if you return something diferent than false.

anyway, remember to use immutable js for your stores in order to know quicky is something has changed with only a single reference comparison.

I hope this was helpful. :)

**Comment by endragor** [comment]
@mercuriete 
Thanks for your comment.

I know that a change is always emitted, unless you return `false`. And that's what my concern is about. I don't want a change to be emitted, unless I explicitly call `setState`

I don't understand the rationale behind the current behaviour.

**Comment by jdlehman** [comment]
I believe the rationale is that alt is not forcing people to use `setState`, though it is recommended.  So someone can just set data directly on the store `this.myData = 1`, which is why emitting changes is the default behavior.  In this case it makes sense because then it does not require the user to explicitly emit the change, though the tradeoff is that they have to explicitly return false if they do not want to emit a change.  `setState` actually was a more recent addition to alt, but we did not change the default behavior for this (and if we did it would be a breaking change and not show up until a major release).

**Comment by jdlehman** [comment]
Also you can use [`preventDefault`](https://github.com/goatslacker/alt/blob/master/docs/createStore.md#storemodelpreventdefault) in the store to do the same thing as returning false:

**Comment by endragor** [comment]
I see. Thanks for the explanation. Are there any plans to make alt configurable to be able to change the behaviour like the one being discussed? We never change Store fields directly, so we don't need the current behaviour.

**Comment by jdlehman** [comment]
I don't see why not, we generally like providing the ability to configure.

A good place to do something like this would be in the [store config](https://github.com/goatslacker/alt/blob/master/docs/createStore.md#storemodelconfig).  Feel free to submit a PR otherwise I will try to take a look at it today or tomorrow and try to get something working.

**Comment by goatslacker** [comment]
We can implement something akin to React' shouldUpdate: `shouldStoreUpdate` method in the store config. I can call that before triggering an event.

**Comment by jdlehman** [comment]
:+1:  to that naming

**Comment by endragor** [comment]
How can I use that to return `false` whenever setState was not called? Should I additionally override setState to set some flag? Where should I reset the flag?

And are you sure "update" is a good word? In React update = rerender, but here it means emitting change event (which results in update of the views, yes, but nothing happens with the store itself). `shouldEmitChange` would sound more clear to me.

**Comment by jdlehman** [comment]
Right that might be a better name for it. The PR is just a first pass so we can figure out exactly what we want.

You would just need to return false from this function in order to return false whenever `setState` was not called.

**Comment by endragor** [comment]
Not sure I understand it correctly.
As I understand, if I constantly return `false` from the function, then `emitChange` will never be called, even when `setState` was called, right?

**Comment by jdlehman** [comment]
I don't believe so, it should still call `emitChange` when `setState` is called (but this is why I need to add tests so I can confirm).  Feel free to take a stab at it as well.
